# TCC PUC-Minas Gastão Jucá Filho <br>
Turma Puc-Minas Virtual - nov2019 <br>


# Resto da limpeza dos arquivos da Olist

#  Atividades previstas:
> 1. Continuar o notebook "Tarefas até 19abr2021". 
> 2. "Limpar" as tabelas restantes da base da Olist. As tabelas a limpar são:
>> <font color=red>"customers"</font>, <font color=red>"sellers"</font> e <font color=red>"geolocation"</font>;  
> 3. Fazer merges. 
> 4. Documentar tudo com linguagem Markdown.

In [1]:
import pandas as pd
import numpy as np
pd.options.display.float_format = '{:,.2f}'.format
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
#pd.options.display.max_columns = 100
from IPython.display import Image
from IPython.core.display import HTML 

## Importação das tabelas 

In [2]:
# reviews = pd.read_csv("olist_order_reviews_dataset.csv", sep = ',')

In [3]:
# orders = pd.read_csv('olist_orders_dataset.csv', sep=',')

In [4]:
# itens = pd.read_csv("olist_order_items_dataset.csv", sep = ',')

In [5]:
products = pd.read_csv("olist_products_dataset.csv", sep = ',')

In [6]:
customers = pd.read_csv("olist_customers_dataset.csv", sep = ',')

In [7]:
geolocation = pd.read_csv("olist_geolocation_dataset.csv", sep = ',')

In [8]:
# payments = pd.read_csv("olist_order_payments_dataset.csv", sep = ',')

In [9]:
sellers = pd.read_csv("olist_sellers_dataset.csv", sep = ',')

### 4.3 - Tabela <font color=red>"customers"
    > Itemização continua a do notebook anterior

In [10]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [11]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
customer_id                 99441 non-null object
customer_unique_id          99441 non-null object
customer_zip_code_prefix    99441 non-null int64
customer_city               99441 non-null object
customer_state              99441 non-null object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [12]:
# valores únicos dos campos
customers[['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']].nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

#### <font color=magenta>RESUMO</font>. Esses dois primeiros campos podem suscitar dúvidas, vamos esclarecer:
> <font color=green>'customer_id'</font> é a chave primária da tabela <font color=red>"customers"</font>
>> Também é único na tabela <font color=red>"orders"</font>, ou seja, cada compra tem seu 'customer_id'. Acho essa modelagem boboca, mas é assim que está.

> <font color=green>'Customer_unique_id'</font> é único para a cada pessoa. 
>> Se uma pessoa fizer duas compras na Olist, em cada uma, receberá um <font color=green>'customer_id'</font> diferente. Gostaram?

> Não há o que limpar, a tabela <font color=red>"customers"</font> está pronta para uso.

In [13]:
# Nossa sexta tabela se aglutina
Image (url = "seis tabelas1.bmp")

### 4.4 - Tabela <font color=red>"sellers"</font>

In [1]:
sellers.head()

NameError: name 'sellers' is not defined

In [15]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
seller_id                 3095 non-null object
seller_zip_code_prefix    3095 non-null int64
seller_city               3095 non-null object
seller_state              3095 non-null object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [16]:
# valores únicos dos campos
sellers[['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']].nunique()

seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

#### <font color=magenta>RESUMO</font>. Fornecedores que enviam os produtos aos clientes.
> <font color=green>'seller_id'</font> é a chave primária da tabela <font color=red>"sellers"</font>

> Não há o que limpar, a tabela <font color=red>"sellers"</font> está pronta para uso.

In [17]:
# Sétima tabela surge
Image (url = "sete tabelas.bmp")

#### 4.4.1 - Adição do campo <font color=green>'business_segment'</font> à tabela <font color=red>"sellers"</font>

##### 4.4.1a - A partir da tabela <font color=red>"closed_deals"</font>
> A informação poderá ser cruzada com 'product_category_name' da tabela <font color=red>"products"</font>

In [18]:
# Leitura da tabela "closed deals"
closed_deals = pd.read_csv("olist_closed_deals_dataset.csv", sep = ',')

In [19]:
closed_deals.head()

mql_id                         seller_id  \
0  5420aad7fec3549a85876ba1c529bd84  2c43fb513632d29b3b58df74816f1b06   
1  a555fb36b9368110ede0f043dfc3b9a0  bbb7d7893a450660432ea6652310ebb7   
2  327174d3648a2d047e8940d7d15204ca  612170e34b97004b3ba37eae81836b4c   
3  f5fee8f7da74f4887f5bcae2bafb6dd6  21e1781e36faf92725dde4730a88ca0f   
4  ffe640179b554e295c167a2f6be528e0  ed8cb7b190ceb6067227478e48cf8dde   

                             sdr_id                             sr_id  \
0  a8387c01a09e99ce014107505b92388c  4ef15afb4b2723d8f3d81e51ec7afefe   
1  09285259593c61296eef10c734121d5b  d3d1e91a157ea7f90548eef82f1955e3   
2  b90f87164b5f8c2cfa5c8572834dbe3f  6565aa9ce3178a5caf6171827af3a9ba   
3  56bf83c4bb35763a51c2baab501b4c67  d3d1e91a157ea7f90548eef82f1955e3   
4  4b339f9567d060bcea4f5136b9f5949e  d3d1e91a157ea7f90548eef82f1955e3   

              won_date business_segment      lead_type lead_behaviour_profile  \
0  2018-02-26 19:58:54              pet  online_medium                    cat   
1  2018-05-08 20:17:59  car_accessories       industry                  eagle   
2  2018-06-05 17:27:23  home_appliances     online_big                    cat   
3  2018-01-17 13:51:03       food_drink   online_small                    NaN   
4  2018-07-03 20:17:45  home_appliances       industry                   wolf   

  has_company has_gtin average_stock business_type  \
0         NaN      NaN           NaN      reseller   
1         NaN      NaN           NaN      reseller   
2         NaN      NaN           NaN      reseller   
3         NaN      NaN           NaN      reseller   
4         NaN      NaN           NaN  manufacturer   

   declared_product_catalog_size  declared_monthly_revenue  
0                            nan                      0.00  
1                            nan                      0.00  
2                            nan                      0.00  
3                            nan                      0.00  
4                            nan                      0.00

In [20]:
closed_deals.shape

(842, 14)

In [21]:
closed_deals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
mql_id                           842 non-null object
seller_id                        842 non-null object
sdr_id                           842 non-null object
sr_id                            842 non-null object
won_date                         842 non-null object
business_segment                 841 non-null object
lead_type                        836 non-null object
lead_behaviour_profile           665 non-null object
has_company                      63 non-null object
has_gtin                         64 non-null object
average_stock                    66 non-null object
business_type                    832 non-null object
declared_product_catalog_size    69 non-null float64
declared_monthly_revenue         842 non-null float64
dtypes: float64(2), object(12)
memory usage: 92.2+ KB


###### WHAT????? Only 842 'seller_id'?????
> Mas há 3.095 na tabela <font color=red>"sellers"</font>!!!!!

> Alguém me passe um WhatsApp, explicando, please...

In [22]:
# Por que só há 842 'seller_id' na tabela "closed_deals" (tabela da direita) ????? Deveria haver 3.095, como na da esquerda.
Image (url = "sellers e closed_deals.bmp")

In [23]:
# Quantos valores únicos há em 'seller_id'?
closed_deals['seller_id'].nunique()

842

In [24]:
# Quantos valores únicos há em 'business_segment'?
closed_deals['business_segment'].nunique()

33

In [25]:
# Quais são eles?
closed_deals['business_segment'].unique()

array(['pet', 'car_accessories', 'home_appliances', 'food_drink',
       'health_beauty', 'computers', 'household_utilities',
       'construction_tools_house_garden', 'toys', 'sports_leisure',
       'stationery', 'food_supplement', 'home_decor', 'bed_bath_table',
       'watches', 'fashion_accessories', 'jewerly', 'party',
       'small_appliances', 'audio_video_electronics', 'other',
       'bags_backpacks', 'home_office_furniture', 'music_instruments',
       'books', 'baby', 'air_conditioning', 'phone_mobile', nan,
       'handcrafted', 'perfume', 'gifts', 'religious', 'games_consoles'],
      dtype=object)

###### Repetindo: a ideia é trazer 'business_segment' (da tabela <font color=red>"closed_deals"</font>) para a tabela <font color=red>"sellers"</font>

##### 4.4.1b - Antes, é preciso garantir descrições idênticas às de 'product_category_name"

In [26]:
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                40.00                      287.00                1.00   
1                44.00                      276.00                1.00   
2                46.00                      250.00                1.00   
3                27.00                      261.00                1.00   
4                37.00                      402.00                4.00   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0            225.00              16.00              10.00             14.00  
1          1,000.00              30.00              18.00             20.00  
2            154.00              18.00               9.00             15.00  
3            371.00              26.00               4.00             26.00  
4            625.00              20.00              17.00             13.00

In [27]:
# valores únicos de 'product_category_name'
df_categorias_unicas = pd.DataFrame(products['product_category_name'].unique(), columns=['product_category_name'])
df_categorias_unicas.sort_values(by='product_category_name').head(10)

product_category_name
29  agro_industria_e_comercio
55                  alimentos
53          alimentos_bebidas
1                       artes
63         artes_e_artesanato
45          artigos_de_festas
60           artigos_de_natal
52                      audio
17                 automotivo
3                       bebes

###### Como a descrição acima está em português, precisaremos da tradução de 'business_segment'
> Que obteremos na tabela <font color=red>"category_translation"</font>

In [28]:
# Leitura da tabela "category_translation"
category_translation = pd.read_csv("product_category_name_translation.csv", sep = ',')

In [29]:
category_translation.sort_values(by='product_category_name').head(10)

product_category_name product_category_name_english
37  agro_industria_e_comercio    agro_industry_and_commerce
45                  alimentos                          food
10          alimentos_bebidas                    food_drink
46                      artes                           art
67         artes_e_artesanato         arts_and_craftmanship
34          artigos_de_festas                party_supplies
54           artigos_de_natal            christmas_supplies
20                      audio                         audio
2                  automotivo                          auto
11                      bebes                          baby

In [30]:
category_translation.shape

(71, 2)

##### As informações em língua portuguesa são idênticas (pelo menos as 10 primeiras em ordem alfabética). 
> Vamos inserir a informação, usando o comando 'merge'

In [31]:
# exemplo de merge  --->    m = pd.merge(tabela_1, tabela_2, how = 'left', on = 'Nome')
sellers_com_bSegment = pd.merge(sellers, closed_deals, how = 'left', on = 'seller_id')

In [32]:
sellers_com_bSegment.shape

(3095, 17)

In [33]:
sellers_com_bSegment.head(5)

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state mql_id sdr_id sr_id won_date  \
0           campinas           SP    NaN    NaN   NaN      NaN   
1         mogi guacu           SP    NaN    NaN   NaN      NaN   
2     rio de janeiro           RJ    NaN    NaN   NaN      NaN   
3          sao paulo           SP    NaN    NaN   NaN      NaN   
4  braganca paulista           SP    NaN    NaN   NaN      NaN   

  business_segment lead_type lead_behaviour_profile has_company has_gtin  \
0              NaN       NaN                    NaN         NaN      NaN   
1              NaN       NaN                    NaN         NaN      NaN   
2              NaN       NaN                    NaN         NaN      NaN   
3              NaN       NaN                    NaN         NaN      NaN   
4              NaN       NaN                    NaN         NaN      NaN   

  average_stock business_type  declared_product_catalog_size  \
0           NaN           NaN                            nan   
1           NaN           NaN                            nan   
2           NaN           NaN                            nan   
3           NaN           NaN                            nan   
4           NaN           NaN                            nan   

   declared_monthly_revenue  
0                       nan  
1                       nan  
2                       nan  
3                       nan  
4                       nan

In [34]:
# Deleção dos campos não mais necessários
sellers_normalizado = sellers_com_bSegment.drop(['mql_id', 'sdr_id','sr_id', 'won_date', 'lead_type', 'lead_behaviour_profile', 'has_company', 'has_gtin', 'average_stock', 'business_type', 'declared_product_catalog_size', 'declared_monthly_revenue'], axis=1)
sellers_normalizado.tail()

seller_id  seller_zip_code_prefix seller_city  \
3090  98dddbc4601dd4443ca174359b237166                   87111     sarandi   
3091  f8201cab383e484733266d1906e2fdfa                   88137     palhoca   
3092  74871d19219c7d518d0090283e03c137                    4650   sao paulo   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080     pelotas   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051     taubate   

     seller_state business_segment  
3090           PR              NaN  
3091           SC              NaN  
3092           SP              NaN  
3093           RS              NaN  
3094           SP              NaN

In [35]:
sellers_normalizado.shape

(3095, 5)

In [36]:
sellers_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3095 entries, 0 to 3094
Data columns (total 5 columns):
seller_id                 3095 non-null object
seller_zip_code_prefix    3095 non-null int64
seller_city               3095 non-null object
seller_state              3095 non-null object
business_segment          380 non-null object
dtypes: int64(1), object(4)
memory usage: 145.1+ KB


In [37]:
# Quantos valores únicos há em 'business_segment'?
sellers_normalizado['business_segment'].nunique()

29

###### Somente 380 'business_segment' em 3.090 'seller_id'.
> É outra incongruência desse DataSet.
>> Provavelmente falha no carregamento de dados.

In [38]:
# Quantos valores únicos há em 'business_segment'?
sellers_normalizado['business_segment'].nunique()

29

In [39]:
# Poderemos relacionar 'product_category_name' (da tabela "products") à 'business_segment' (da tabela "sellers_normalizada")
Image (url = "sete tabelas com sellers_normalizada.bmp")

### 4.5 - Tabela <font color=red>"geolocation"</font>

In [40]:
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037           -23.55           -46.64   
1                         1046           -23.55           -46.64   
2                         1046           -23.55           -46.64   
3                         1041           -23.54           -46.64   
4                         1035           -23.54           -46.64   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [41]:
geolocation.shape

(1000163, 5)

In [42]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
geolocation_zip_code_prefix    1000163 non-null int64
geolocation_lat                1000163 non-null float64
geolocation_lng                1000163 non-null float64
geolocation_city               1000163 non-null object
geolocation_state              1000163 non-null object
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [43]:
# valores únicos dos campos
geolocation[['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']].nunique()

geolocation_zip_code_prefix     19015
geolocation_lat                717358
geolocation_lng                717613
geolocation_city                 8011
geolocation_state                  27
dtype: int64

#### 4.5.1 - Busca pela chave primária da tabela <font color=red>"geolocation"</font>

In [44]:
# Qual a chave primária da tabela "geolocation"?
# Vamos deletar duplicados da concatenação de 'geolocation_zip_code_prefix' com 'geolocation_lat' (Se não houver deleção, eis a chave primária)
len(geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat']))

719298

###### Não era aquela, vamos tentar outra.

In [45]:
len(geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lng']))

719445

###### Também não era aquela.

In [46]:
len(geolocation.drop_duplicates (subset=['geolocation_lat', 'geolocation_lng']))

718463

###### Nem aquela

In [47]:
len(geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng']))

720154

###### ???

In [48]:
len(geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city']))

738330

In [49]:
len(geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']))

738332

###### Não há chave primária na tabela?????
> Olhemos com lupa a coluna 'geolocation_zip_code_prefix'

In [50]:
# Listagem da geolocation que contém o 'geolocation_zip_code_prefix' 1046
geolocation.loc[geolocation['geolocation_zip_code_prefix'] == 1046].sort_values(by='geolocation_city')

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
1                            1046           -23.55           -46.64   
996                          1046           -23.55           -46.64   
981                          1046           -23.55           -46.64   
971                          1046           -23.55           -46.64   
970                          1046           -23.55           -46.64   
932                          1046           -23.55           -46.64   
923                          1046           -23.55           -46.64   
914                          1046           -23.55           -46.64   
899                          1046           -23.55           -46.64   
893                          1046           -23.55           -46.64   
857                          1046           -23.55           -46.64   
850                          1046           -23.55           -46.64   
839                          1046           -23.55           -46.65   
828                          1046           -23.55           -46.64   
1011                         1046           -23.55           -46.65   
822                          1046           -23.55           -46.64   
806                          1046           -23.55           -46.64   
793                          1046           -23.55           -46.64   
778                          1046           -23.55           -46.64   
762                          1046           -23.55           -46.64   
756                          1046           -23.55           -46.64   
742                          1046           -23.55           -46.65   
740                          1046           -23.55           -46.64   
689                          1046           -23.55           -46.64   
687                          1046           -23.55           -46.65   
1450                         1046           -23.55           -46.64   
675                          1046           -23.55           -46.64   
656                          1046           -23.55           -46.64   
650                          1046           -23.55           -46.65   
813                          1046           -23.55           -46.64   
1037                         1046           -23.55           -46.64   
1039                         1046           -23.55           -46.64   
1053                         1046           -23.55           -46.64   
1439                         1046           -23.55           -46.64   
1437                         1046           -23.55           -46.64   
1432                         1046           -23.55           -46.64   
1428                         1046           -23.55           -46.64   
1414                         1046           -23.55           -46.64   
1412                         1046           -23.55           -46.64   
1404                         1046           -23.55           -46.64   
1387                         1046           -23.55           -46.64   
1375                         1046           -23.55           -46.64   
1373                         1046           -23.55           -46.64   
1372                         1046           -23.55           -46.64   
1367                         1046           -23.55           -46.64   
1361                         1046           -23.55           -46.64   
1358                         1046           -23.55           -46.65   
1356                         1046           -23.55           -46.64   
1341                         1046           -23.55           -46.64   
1311                         1046           -23.55           -46.64   
1310                         1046           -23.55           -46.64   
1305                         1046           -23.55           -46.64   
1292                         1046           -23.55           -46.64   
1275                         1046           -23.55           -46.64   
1242                         1046           -23.55           -46.64   
1206                         1

In [51]:
# Quantos registros são?
len(geolocation.loc[geolocation['geolocation_zip_code_prefix'] == 1046].sort_values(by='geolocation_city'))

141

###### Não é à toa que não achava a chave primária! Há muitos duplicados!!!!!!
> Precisamos eliminá-los.

In [52]:
geolocation_sem_repetir = geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state'])

#### 4.5.2 - Busca pela chave primária da tabela <font color=red>"geolocation_sem_repetir"</font>

In [53]:
geolocation_sem_repetir.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037           -23.55           -46.64   
1                         1046           -23.55           -46.64   
2                         1046           -23.55           -46.64   
3                         1041           -23.54           -46.64   
4                         1035           -23.54           -46.64   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [54]:
geolocation_sem_repetir.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 738332 entries, 0 to 1000161
Data columns (total 5 columns):
geolocation_zip_code_prefix    738332 non-null int64
geolocation_lat                738332 non-null float64
geolocation_lng                738332 non-null float64
geolocation_city               738332 non-null object
geolocation_state              738332 non-null object
dtypes: float64(2), int64(1), object(2)
memory usage: 33.8+ MB


In [55]:
geolocation_sem_repetir.shape

(738332, 5)

In [56]:
# valores únicos dos campos
geolocation_sem_repetir[['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']].nunique()

geolocation_zip_code_prefix     19015
geolocation_lat                717358
geolocation_lng                717613
geolocation_city                 8011
geolocation_state                  27
dtype: int64

In [57]:
# Qual a chave primária da tabela "geolocation_sem_repetir"?
# Vamos deletar duplicados da concatenação de 'geolocation_zip_code_prefix' com 'geolocation_lat' (Se não houver deleção, eis a chave primária)
len(geolocation_sem_repetir.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat']))

719298

In [58]:
len(geolocation_sem_repetir.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lng']))

719445

In [59]:
len(geolocation_sem_repetir.drop_duplicates (subset=['geolocation_lat', 'geolocation_lng']))

718463

In [60]:
len(geolocation_sem_repetir.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng']))

720154

In [61]:
len(geolocation_sem_repetir.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city']))

738330

In [62]:
len(geolocation_sem_repetir.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']))

738332

##### Cacilda!
> A chave primária é a concatenação de TODOS os campos da tabela.

##### A nova configuração de tabelas chega quase ao final!

In [63]:
Image (url = "Tabelas quase completas.PNG")

##### "Quase" ao final porque faltam duas coisas:
> 1. Checar se cabe outra solução que não deleção de itens repetidos na tabela <font color = red>"itens"</font>.
> 2. Juntar todos as colunas relavantes das 8 tabelas acima em uma só.

> Isso será feito em notebook específico. Este chega ao final aqui.